onnxruntime/core/mlas/lib/sqnbitgemm_kernel_neon.cpp

```diff
@@ -0,0 +1,515 @@
+}
+
+template <size_t Capacity>
+MLAS_FORCEINLINE std::array<float32x4_t, Capacity / 4>
+LoadDataAndZeroPad(const float* data, size_t count)
+{
+    static_assert(Capacity % 4 == 0, "Capacity must be divisible by 4.");
+
+    count = std::min(count, Capacity);
+
+    std::array<float32x4_t, Capacity / 4> dst{};
+
+    size_t vi = 0;  // vector index
+
+    // handle 4 values at a time
+    while (count > 3) {
+        dst[vi] = vld1q_f32(data);
+
+        vi += 1;
+        data += 4;
+        count -= 4;
+    }
+
+    // handle remaining values
+    if (count > 0) {
+        dst[vi] = vsetq_lane_f32(data[0], dst[vi], 0);
+    }
+
+    if (count > 1) {
+        dst[vi] = vsetq_lane_f32(data[1], dst[vi], 1);
+    }
+
+    if (count > 2) {
+        dst[vi] = vsetq_lane_f32(data[2], dst[vi], 2);
+    }
+
+    return dst;
+}
+
+template <size_t BlkBitWidth, size_t BlkLen, size_t NCols>
+MLAS_FORCEINLINE void
+ComputeDotProducts(
+    const float* ARowPtr,
+    const uint8_t* QuantBDataColPtr,
+    const float* QuantBScaleColPtr,
+    const uint8_t* QuantBZeroPointColPtr,
+    float* SumPtr,
+    size_t CountK,
+    size_t StrideQuantBData,
+    size_t StrideQuantBScale,
+    size_t StrideQuantBZeroPoint,
+    const float* BiasPtr
+)
+{
+    static_assert(NCols == 1 || NCols == 4, "NCols must be 1 or 4");
+
+    const uint8x8_t LowMask = vdup_n_u8(0x0F);
+
+    float32x4_t acc[NCols]{};
+
+    const uint8_t* QuantBData = QuantBDataColPtr;
+    const float* QuantBScale = QuantBScaleColPtr;
+    size_t QuantBZeroPointIdx = 0;  // track half byte increments with this index instead of a pointer
+
+    for (size_t k = 0; k < CountK; k += BlkLen) {
+        const size_t k_blk_len = std::min(CountK - k, BlkLen);
+
+        const uint8_t* b_data[NCols];
+        UnrolledLoop<NCols>(
+            [&](size_t i) { b_data[i] = QuantBData + i * StrideQuantBData; }
+        );
+
+        float scale[NCols];
+        UnrolledLoop<NCols>(
+            [&](size_t i) { scale[i] = QuantBScale[i * StrideQuantBScale]; }
+        );
+
+        float offset[NCols];  // Includes zero point and float conversion offset of 16.
+                              // More details on the float conversion below.
+        if (QuantBZeroPointColPtr != nullptr) {
+            UnrolledLoop<NCols>([&](size_t i) {
+                const uint8_t zp_packed =
+                    QuantBZeroPointColPtr[i * StrideQuantBZeroPoint + QuantBZeroPointIdx / 2];
+                const uint8_t zp = ((QuantBZeroPointIdx & 1) == 1) ? (zp_packed >> 4) : (zp_packed & 0x0F);
+                offset[i] = 16.0f + zp;
+            });
+        } else {
+            UnrolledLoop<NCols>([&](size_t i) {
+                constexpr uint8_t zp = 8;
+                offset[i] = 16.0f + zp;
+            });
+        }
+
+        constexpr size_t SubBlkLen = 16;  // number of block elements to process in one iteration
+
+        for (size_t k_idx_in_blk = 0; k_idx_in_blk < k_blk_len; k_idx_in_blk += SubBlkLen) {
```

Comment:
> { [](http://example.com/codeflow?start=90&length=2)

you can handle the compute case firstly, and then handle the reminder, just like what you are doing for CountN:

int64_t nblk = static_cast<int64_t>(CountN) - NCols;

while (nblk >= 0) {

Reply:
in this way, we don't need to run the reminder in practice case, like for llama model

Reply:
tried a simple test of removing the remainder handling entirely (i.e., padding data with zeros) to see the potential performance gain.

```c++
            //const size_t k_subblk_len = std::min(k_blk_len - k_idx_in_blk, SubBlkLen);
            //const auto av = LoadDataAndZeroPad<SubBlkLen>(ARowPtr + k + k_idx_in_blk, k_subblk_len);
            const float* a_data = ARowPtr + k + k_idx_in_blk;
            const float32x4_t av[4] = {
                vld1q_f32(a_data),
                vld1q_f32(a_data + 4),
                vld1q_f32(a_data + 8),
                vld1q_f32(a_data + 12),
            };
```

there was not a significant improvement.

I think there are other bigger optimizations to try first. perhaps we can revisit this (which probably means some code duplication) after optimizing the code more.